Envoy-based ingress controller

zcahana opened the pull request:
This is a first-pass on Envoy-based ingress controller (addressing #6).
I'm PR'ing this while still WIP to start gathering feedback and set this in the right direction.

What we have so far:
- Ingress controller running in the context of the manager (via `manager ingress` command).
- Extending the `platform/kube/Controller` to sync ingress resources.
- Adding ingress watcher that reconfigures Envoy based on the given ingresses.
- Envoy clusters made up of single upstream (service's DNS name).

What's still missing:
- Represent ingress rules in the Istio model, so that proxy/envoy isn't coupled with k8s objects.
- Support for path regex in ingress (under the limitations of what Envoy supports)
- Constructs envoy clusters using SDS + pod endpoints.
- Update ingress status with the proxy's "public" IP address (node IP?).
- TLS termination.
- Testing.

rshriram commented:
@zcahana could you please sign the cla?

zcahana commented:
@rshriram I applied to join the ibm-contributors group, waiting for it to get approved.

rshriram commented on cmd/manager/main.go:
```diff
@@ -122,6 +138,18 @@ func init() {
+		"Discovery service DNS address")
+	ingressCommand.PersistentFlags().IntVarP(&flags.proxy.AdminPort, "admin_port", "a", 5000,
+		"Envoy admin port")
+	ingressCommand.PersistentFlags().StringVarP(&flags.proxy.BinaryPath, "envoy_path", "b", "/envoy_esp",
```
Envoy's admin port is of little use to end user.  It is a readonly port. I don't even think this needs to be exposed.
Could you also add a TODO here saying that the hardcoded use of envoy must be changed to something more generic? e.g., 
``//TODO: Change to proxy_type: envoy|nginx|haproxy, and then make the rest (SDS, etc) as sub flags```

rshriram commented on model/BUILD:
```diff
@@ -4,6 +4,7 @@ go_library(
     srcs = [
         "controller.go",
+        "ingress.go",
         "registry.go",
```
Please remove ingress from the model. Model is platform agnostic representation of resources. Mesos may not even have ingress.

zcahana replied:
So, we need something in the model (`RoutingRule`?) that the envoy ingress watcher can consume, and the kube ingresses converted to. Is there already any work in that direction?

rshriram replied:
+1. Model package is completely platform independent.

rshriram commented on model/controller.go:
```diff
@@ -34,6 +36,9 @@ type Controller interface {
 
+	// AppendIngressHandler notifies about changes to ingress resources
+	AppendIngressHandler(f func(*v1beta1.Ingress, Event)) error
+
```
May be add a type to the Config struct to detect if its an ingress or not. This way, ingress becomes kube specific.

zcahana replied:
I'm not sure I understand how the `Config` is designed to be worked with. Any reference or explanation can help.

zcahana replied:
The ingress config seems to be a subset of the model's [`RouteRule`](https://github.com/istio/manager/blob/master/model/proxy/alphav1/config/cfg.pb.go#L153) (that's what you mean by `ProxyConfig`, right)? So, we can have the controller convert ingresses to such rules, and have the proxy watcher subscribe to these.

rshriram replied:
Yes. Please take a look at the Generate function in proxy/envoy/config.go. You can provide an empty array for upstreams and rules. It should give you an envoy configuration for services as hostnames and / as the default prefix. Along with SDS functionality. In next iteration (or if you want to hack the utility functions), you could generate Envoy route entries under a virtual host with specific path or prefixes. 

Bottom line. There is opportunity for code reuse. The config.go takes care of generating all the right listener blocks (with use_orig_dst, etc).

rshriram commented on model/ingress.go:
```diff
@@ -0,0 +1,23 @@
+type IngressRegistry interface {
+	// ListIngresses all ingress rules currently defined.
+	ListIngresses() []*v1beta1.Ingress
+}
```
Please (re)move was you see fit.

rshriram commented on proxy/envoy/ingress.go:
```diff
@@ -0,0 +1,244 @@
+type ingressWatcher struct {
+	agent   Agent
+	mesh    *MeshConfig
+	ingress model.IngressRegistry
```
one favor: could you add a place holder or a TODO, so that #32 can be incorporated here easily as well?

zcahana replied:
Not sure I understand that comment...

zcahana replied:
I'd suggest to postpone merging the two watchers until we get a clearer picture of how much logic they share. At least for now, they are generating quite a different envoy configuration. The sidecar watcher shouldn't care / handle ingress-based configurations, whereas the ingress watcher shouldn't allow non-ingressed traffic.

kyessenov replied:
Sounds good to me. We should share code for config generation but the controllers themselves should be independent of one another. Let's see how it works out and extract proper module interfaces later.

rshriram commented on proxy/envoy/ingress.go:
```diff
@@ -0,0 +1,244 @@
+	time.Sleep(256 * time.Millisecond)
+}
+
+func (w *ingressWatcher) generateConfig() (*Config, error) {
```
is there a way to de-duplicate this functionality by moving code out of proxy/envoy/watcher.go(?) that generates Envoy configuration, and reusing it in both places? After all, that code just expects a list of services, pods, and rules. in the absence of rules, it behaves just like a user-space kube proxy (except operating on per-pod basis). Whatever is missing will not be configured.

rshriram replied:
On second thought, lets add a TODO here as a reminder that we need to refactor this code. We could do that in a later PR.
Kind of similar to the IdentityProvider stuff in A8, in principle. What do you think?

zcahana replied:
Yes, there's some de-duplication and code reuse that can be done. Will be refactored in next iterations on this.

rshriram commented on proxy/envoy/ingress.go:
```diff
@@ -0,0 +1,244 @@
+		}
+
+		// TODO: Validate no duplicate path/routes within this virtualhost
+		// (e.g., due to 2 default backends defined in 2 ingresses in different namespaces)
```
silly question: is ingress common to all namespaces? or specific to a namespace? IF latter, we don't need this right?

zcahana replied:
Ingresses are defined within namespaces, but are normally realized by a single namespace-neutral ingress controller, which serves incoming traffic to all namespaces. This TODO refers to the need to make sure there are no collision between ingresses in different namespaces, which might yield invalid envoy configuration (e.g., a virtualhost with 2 similar routes, or 2 prefix-similar routes, etc).

rshriram commented on proxy/envoy/ingress.go:
```diff
@@ -0,0 +1,244 @@
+	listeners = append(listeners, Listener{
+		Port:           80,
+		BindToPort:     true,
+		UseOriginalDst: true,
```
Set to false please. This does not apply to ingress. Its only for internal envoy instances.

rshriram commented on proxy/envoy/ingress.go:
```diff
@@ -0,0 +1,244 @@
+				Type:             "strict_dns", // TODO: Change to SDS
+				LbType:           DefaultLbType,
+				ConnectTimeoutMs: DefaultTimeoutMs,
+				Hosts:            clusterHosts(namespace, backend),
```
don't we have to set the cluster `"features": "http2"`, since we would have Envoy backends? If this is for istio only, then by default (and for first cut) I think we should set to features to "http2", since all http communication happens over Envoy to Envoy. If not specified, Envoy falls back to http/1.1, nullifying a lot of benefits that we get out of http/2.

zcahana replied:
I think this is Istio-only for now. Will set "http2".

kyessenov replied:
I'll try to refactor current config to make it more re-usable. It handles things like enabling http2.

rshriram commented on proxy/envoy/ingress.go:
```diff
@@ -0,0 +1,244 @@
+		Listeners: listeners,
+		ClusterManager: ClusterManager{
+			Clusters: clusters,
+		},
```
please set the `"local_cluster_name": "ingress"`. Local cluster name is needed for a variety of features in envoy (CDS, tracing, possibly RDS, etc.).

zcahana replied:
Reading [here](https://lyft.github.io/envoy/docs/configuration/cluster_manager/cluster_manager.html), it seems that `local_cluster_name` is only needed for enabling zone aware routing. This is probably out of the scope of this PR.

kyessenov commented on proxy/envoy/ingress.go:
```diff
@@ -0,0 +1,244 @@
+	clustersMap := make(map[string]Cluster)
+
+	// Function to add a rule for a particular host, path and backend.
+	// This will later be invoked for each ingress rule, as well as ingress default backend
```
We need to provide a service for a 404 backend. The service can be implemented by Envoy or default-http-backend from kubernetes.

zcahana replied:
This functionality is already exposed by k8s' ingresses, which allows the user to specify a default backend (service) for requests that don't match any ingress rule. This is implemented as an Envoy VirtualHost with a wildcard domain ("*"). If the user doesn't specify a default backend in its ingresses, then envoy we don't produce such wildcard VirtualHost, and fallback to Envoy's own 404.

kyessenov commented on proxy/envoy/ingress.go:
```diff
@@ -0,0 +1,244 @@
+	}
+
+	// Iterate on all ingresses, and create corresponding rule
+	for _, ingress := range w.ingress.ListIngresses() {
```
Please make the order of rules or ingresses canonical. Otherwise, the watcher will keep reloading Envoy due to some permutation in the config file.

kyessenov commented:
More comments for the TODOs:
- Constructs envoy clusters using SDS + pod endpoints.
Yes, we should avoid using kube DNS since we have SDS. Don't worry about CDS for now, it's easy to refactor for that later.
- Update ingress status with the proxy's "public" IP address (node IP?).
GCE LB ingress does this because it manages firewall rules and IP allocation. NGINX ingress does not AFAIK and you have to create a separate service of type `LoadBalancerIP` pointing to NGINX deployment. Let's do what NGINX ingress does.
- TLS termination.
Don't worry about it. We will do one big swoop to enable TLS termination.
- Testing.
Take a look at `test/integration/driver.go`. I think it has all the pieces to reuse.

zcahana commented on cmd/manager/main.go:
```diff
@@ -122,6 +138,18 @@ func init() {
+		"Envoy binary location")
+	ingressCommand.PersistentFlags().StringVarP(&flags.proxy.ConfigPath, "config_path", "e", "/etc/envoy",
+		"Envoy config root location")
+	rootCmd.AddCommand(ingressCommand)
```
Yes, but should ingress be a subcommand of proxy? In essence, these are two different modes of operation (in-cluster service sidecar proxy vs. edge ingress proxy).

Most (all?) of the commonalities stem from the fact that both commands manage envoy, so maybe both should be a child command of an "envoy" command?

kyessenov replied:
Yeah, good point. I think we need a shared root, and then run "sidecar", "ingress", and "egress" as a subcommand. Actually, the flag parameter need to be turned into ProxyMeshConfig in the proto, but that's for later.

zcahana commented:
I've rewritten the ingress controller implementation such that Ingress API objects are converted to Istio model rules (currently, same `config.RouteRule`, but with a different kind - `IngressRule`), which are then transformed to Envoy configuration. There's still only a little re-use of existing Envoy configuration code. This can be refactored later when the package stabilizes and the common parts become clearer.

Note that due to the major changes that already went to master, I've rebased my branch and restructured the commits so that they're more reviewable. The commits from the first round are thus gone :).

zcahana commented:
Jenkins' presubmit build quickly fails with the following error ([link](https://istio-testing.appspot.com/job/manager/job/presubmit/270/console)):
```
+ go get ./...
/var/jenkins_home/jobs/manager/jobs/presubmit/workspace/src/main@tmp/durable-3d3f8166/script.sh: 2: /var/jenkins_home/jobs/manager/jobs/presubmit/workspace/src/main@tmp/durable-3d3f8166/script.sh: go: not found
[Pipeline] }
[Pipeline] // dir
[Pipeline] }
[Pipeline] // withEnv
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 127
Finished: FAILURE
```

Not sure what's causing this... any chance it's some recent changes to istio-testing repo? @sebastienvas

sebastienvas commented:
Please rebase your PR.

istio-testing commented:
Jenkins job manager/presubmit started

istio-testing commented:
Jenkins job manager/presubmit failed

istio-testing commented:
Jenkins job manager/presubmit started

istio-testing commented:
Jenkins job manager/presubmit passed

zcahana commented:
thanks @sebastienvas , build now passes.

rshriram commented on platform/kube/conversion.go:
```diff
@@ -147,3 +155,111 @@ func modelToKube(km model.KindMap, k *model.Key, v proto.Message) (*Config, erro
+				// to the proxy configuration generator. This can be improved by using
+				// a dedicated model object for IngressRule (instead of reusing RouteRule),
+				// which exposes the necessary target port field within the "Route" field.
+				Tags: map[string]string{
```
I don't think we do that anymore.

rshriram replied:
I mean refer to ports in the route rule

zcahana replied:
This is necessary for envoy routes which are generated from ingresses - we need to route traffic to the cluster corresponding with the service:port defined by the user in the ingress.

zcahana replied:
Actually, looking further into this, cluster names include the port name, and not the port number.
So, what happens when there's no port name specified?

kyessenov replied:
That can only happen if there is only one port, and the upstream cluster becomes just the service name ("a.svc....") or service-tag combination ("a.svc...::env=prod")

kyessenov replied:
Can you add a TODO item here? This needs to be in the next iteration of routing rules, we've been assuming rules apply per port, but ingress needs an explicit port. Moreover, this is a port name, so this has some implication for the route rule model.

zcahana replied:
This seems to be captured in #151.

zcahana commented:
There's at least one bugfix on the way, and potentially another one. I'm currently running/debugging this locally.

istio-testing commented:
Jenkins job manager/presubmit passed

istio-testing commented:
Jenkins job manager/presubmit started

istio-testing commented:
Jenkins job manager/presubmit passed

istio-testing commented:
Jenkins job manager/presubmit started

istio-testing commented:
Jenkins job manager/presubmit passed

istio-testing commented:
Jenkins job manager/presubmit started

istio-testing commented:
Jenkins job manager/presubmit passed

istio-testing commented:
Jenkins job manager/presubmit passed

zcahana commented on platform/kube/controller.go:
```diff
@@ -298,6 +399,22 @@ func (c *Controller) List(kind, namespace string) (map[model.Key]proto.Message,
 	return out, errs
 }
 
+func (c *Controller) listIngresses(kind, namespace string) (map[model.Key]proto.Message, error) {
```
What is it that's not needed here? The `listIngresses` method gets called from `Controller.List(kind, namespace)` according to the actual kind. In practice, this is being used by the envoy's ingress watcher.

kyessenov replied:
Oops my bad, I was referring to "kind" parameter.

zcahana replied:
Well, you're right that the "kind" parameter isn't really needed here. I left it for uniformity with the other listX methods.

istio-testing commented:
Jenkins job manager/presubmit passed

istio-testing commented:
Jenkins job manager/presubmit passed

zcahana commented:
With the latest round of commits, I've verified that basic ingress function works (i.e., ingress controller forwards ingress traffic to correct backend service according to defined routes). In practice, the backend service's reverse proxy is still blocking the request from hitting the actual service (due to #150), so this is still not usable end-to-end, but probably solid enough for first-pass --> merging.

elevran commented:
Together with #177 - fixes #6
